### DISCUSSION TITLE
Cross Repository Codespaces or Codespaces Compose?

### DISCUSSION THREAD

**Discussion by metaskills** [community/community#5742] [Codespaces]
Many of our smaller & isolated applications (each in a GitHub repository) can easily take advantage of Codespaces. That said, I wonder if y'all might have any advice on how to use Codespaces with 2 or more repositories? For example, if a team has 3 GitHub projects that collectively form an application group, would the recommendation be to create a new `myorg/codespaces-team` repository that would script up checking each of team's needed repositories into the `workspaces` folder, maybe even calling each repo's bootstrap/setup scripts or even their own Codespaces devcontainer hooks? 

This could be a solution for https://github.com/github/feedback/discussions/5741 maybe?

**Comment by jozefvaclavik** [comment]
Check https://github.com/Chuxel/codespaces-multi-repo

**Reply by metaskills** [reply]
Thanks, I also had more details and feedback in this semi-related thread. Will investigate. https://github.com/github/feedback/discussions/5741#discussioncomment-1524872

**Comment by metaskills** [comment]
Thanks, will check that in a bit. Also, note to self. https://code.visualstudio.com/docs/remote/containers#_can-i-connect-to-multiple-containers-at-once

**Comment by metaskills** [comment]
Thanks! These were informative to read and learn, however, I'm not sure it will work for us since it seems to create a kitchen sink that would encompass all the dependencies needed for each project. Our organization has > 100 applications/repositories that are mostly HTTP services. Most of these projects are developed in isolation, meaning each can implement Codespaces and their devcontainer implementation in an isolated fashion.

In some cases, a single service (A) may need work to one or more (B) and (C) HTTP services. If these projects were local, that would be easy to do. Simply change a config in service (A) to point to the local version of (B) and (C). See illustration of what we would need from Codespaces to make this work. Basically, can the SSH tunnel (or whatever mechanics are at play) facilitate port requests back to the Local OS?

![Cross Codespace Requests](https://user-images.githubusercontent.com/2381/147418587-0b31976b-75ad-45e0-ba01-5547841e9433.png)

**Reply by metaskills** [reply]
For example, I noticed a few months ago that Gitpod partnered (https://www.gitpod.io/blog/tailscale) with Tailscale (https://tailscale.com) and wrote up docs on how to use this tool (https://www.gitpod.io/docs/configure/tailscale) to achieve what I am looking for. Is this on the Codespaces team radar?

**Reply by jozefvaclavik** [reply]
I think these are two different issues that look similar.

1. how to clone multiple repositories into same codespace - the only workaround is to create github token and use that instead of default git auth method.
2. how to connect multiple codespaces together - afaik you should be able to install and configure tailscale inside of your Dockerimage.

**Reply by metaskills** [reply]
Yes, I would very muck like to emphasis I am talking about #2 here. It is my hope that since Codespaces team owns the SSH tunnel that allows forwarded ports to work, they have an opinion on this matter too. I think installing Tailscale is one short term solution. But much like native git+ssh possibly coming, I'd like to hear some product roadmap thoughts around this need as well from the team.

**Reply by jozefvaclavik** [reply]
I don't know why you think its short term solution. I'm running my custom image in codespaces that has nothing to do with default image. I prefer to have complete control over the environment that I'm running. I use it on 5+ repositories. Never had any issue with it. Having option to configure it to 100% of your needs is imho way more valuable then having one-fit pre-configured solution.

**Reply by metaskills** [reply]
> I don't know why you think its short term solution.

Hey Jozef, my goal is not to debase any value a solution has for you, myself, or others. Rather, I am providing feedback and asking questions of the Codespaces product team to see if this feature is something that should be more commoditized to add value to myself and others. That feels natural. For example, I could script my own Docker install but the Codespaces product now has "features" to support higher tooling interfaces. Yes we can have escape hatches to 100% control on all sorts of things, but I am interested in a) the team hearing my feedback and b) hearing their thoughts as well. Hope that helps.

**Reply by metaskills** [reply]
https://github.com/github/feedback/discussions/5741#discussioncomment-1883398
